src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -274,11 +276,26 @@ public CompiledScript compileInternal(Script script) {
         CompiledScript compiled = cache.getIfPresent(cacheKey);
         if (compiled == null) {
             //Either an un-cached inline script or an indexed script
-            compiled = new CompiledScript(lang, scriptEngineService.compile(code));
+            try {
+                compiled = new CompiledScript(script.getType(), name, lang, scriptEngineService.compile(code));
+            } catch (RuntimeException re) {
+                throw new RuntimeException("Compilation error with " + script.getType().name()
+                        + "script [" + name + "]: " + re.getMessage(), re);
+            }
             //Since the cache key is the script content itself we don't need to
             //invalidate/check the cache if an indexed script changes.
             cache.put(cacheKey, compiled);
         }
+
+        // An indexed or inline script may be mistaken for the other since the cache key relies on
+        // only the code of the script. Two indexed scripts may also have the same code, but different
+        // names. If this is case, the name and type parameters must be modified accordingly.
+        // Then cache the most recent version used.
+        if (compiled.type() != script.getType() || name != null && !name.equals(compiled.name())) {
```

Comment:
@rmuir Thanks for the feedback -- I don't think this code is going to remain once the cache key is changed, but if it does I will make sure it's simplified in this form.